larpandora/LArPandoraEventBuilding/LArPandoraShower/Tools/Cheating/ShowerTrackFinderCheater_tool.cc

```diff
@@ -122,22 +127,37 @@ namespace ShowerRecoTools {
     ShowerEleHolder.GetElement(fShowerDirectionInputTag,ShowerDirection);
 
     auto const hitHandle = Event.getValidHandle<std::vector<recob::Hit> >(fHitModuleLabel);
-    std::vector<art::Ptr<recob::Hit> > hits;
-    art::fill_ptr_vector(hits, hitHandle);
 
     // Get the hits associated with the space points
     art::FindManyP<recob::SpacePoint> fmsph(hitHandle, Event, fPFParticleLabel);
     if(!fmsph.isValid()){
       throw cet::exception("ShowerTrackFinderCheater") << "Spacepoint and hit association not valid. Stopping.";
     }
 
+    std::vector<int> trueParticleIdVec;
+
+    // If we have an electron, take the hits from the primary only
+    // This will also cover the cases Pandora misclassify a track as a shower
+    if (trueParticle->PdgCode() != 22)
+    {
+      trueParticleIdVec.push_back(trueParticle->TrackId());
+    } else {
+      // If we roll up showers, we have no choice but to take all of the hits from the photon
+      trueParticleIdVec.push_back(-trueParticle->TrackId());
+      // If we do not roll up the showers, take all of the primary daughters
+      for (int i=0; i<trueParticle->NumberDaughters(); i++)
+      {
+        trueParticleIdVec.push_back(trueParticle->Daughter(i));
+      }
+    }
```

Comment:
Can you clarify that I'm following this correctly.
Regardless of whether rollup is enabled or not, you always store the negative ID and the ID of the daughters.
So, if rollup is enabled, the shower has no daughters and the loop over NDaughters does nothing, meaning you only ever store the negative of the primary ID.
If rollup is disabled, you store the negative ID and the primary daughter IDs BUT no hits are matched to a negative ID, so you only ever match hits from the daughters?

Reply:
Yeah, that is exactly the intent here. Now you have pointed it out, I realise that I should probably put this in an if-else block and only save the parent ID if there are no daughters.

Reply:
I updated the logic here to be a bit clearer (and more efficient)